refactor: cleanup notifications system

iamacook opened the pull request:
## What it solves
An introduction to https://github.com/safe-global/safe-pm/issues/83

## How this PR fixes it
The current notification system has been refactored:

- [Notistack](https://github.com/iamhosseindhv/notistack) upgraded to the newest MUI4-supporting version.
- System has been simplified: the new actions are `showNotification`, `close/closeAllNotification(s)` and `delete/deleteAllNotification(s)` and the syntax follows Notistack as closely as possible.
- Immutable has been removed from the `notifications` store.
- Notifications are now cached for later retrieval.

None of the above adjustments should yet reflect/alter the UI in it's current state.

## How to test it
Interact with the Safe and observe notifications functioning as before.

One notable notification to check, which includes an `onClick` handler, is dispatched when the Safe requires an update. I removed some unnecessary code that is responsible for address book interaction notifications. They should still work as before as well.

iamacook commented on src/logic/notifications/store/notifications.ts:
```diff
@@ -0,0 +1,74 @@
+    },
+    [NOTIFICATION_ACTIONS.CLOSE_ALL]: (state) => {
+      return state.map((notification) => ({ ...notification, dismissed: true }))
+    },
+    [NOTIFICATION_ACTIONS.DELETE]: (state, { payload }: Action<KeyPayload>) => {
+      return state.filter((notification) => notification.options?.key !== payload.key)
+    },
+    [NOTIFICATION_ACTIONS.DELETE_ALL]: () => {
+      return []
+    },
+  },
+  [],
+)
+
+export const showNotification = (
+  payload: Pick<Notification, 'message' | 'options'>,
+): ThunkAction<SnackbarKey, AppReduxState, undefined, AnyAction> => {
+  return (dispatch): SnackbarKey => {
+    const action = createAction<Notification>(NOTIFICATION_ACTIONS.SHOW)
+
+    const key = payload.options?.key || Math.random().toString(32).slice(2)
+
+    dispatch(
+      action({
+        ...payload,
+        options: { ...payload.options, key },
+      }),
+    )
+
+    return key
+  }
+}
+export const closeNotification = createAction<KeyPayload>(NOTIFICATION_ACTIONS.CLOSE)
+export const closeAllNotifications = createAction(NOTIFICATION_ACTIONS.CLOSE_ALL)
+export const deleteNotification = createAction<KeyPayload>(NOTIFICATION_ACTIONS.DELETE)
+export const deleteAllNotifications = createAction(NOTIFICATION_ACTIONS.DELETE_ALL)
+
+export const selectNotifications = (state: AppReduxState): NotificationsState => {
+  return state[NOTIFICATIONS_REDUCER_ID]
+}
```
After working on `web-core`, it really doesn't make sense to separate these into separate files in my opinion.

katspaugh commented:
Good refactoring, but it broke some of the notifications. I tested only the address book ones.

<img width="400" alt="Screenshot 2022-06-07 at 10 41 57" src="https://user-images.githubusercontent.com/381895/172336762-ddb9b59d-1f5c-496e-bdf2-94370d0f7042.png">

katspaugh reviewed (changes requested):
Also something is wrong with error styles.

<img width="393" alt="Screenshot 2022-06-07 at 11 20 16" src="https://user-images.githubusercontent.com/381895/172344910-cc21656e-ac78-4d75-982a-4bb8ed49aad2.png">

iamacook commented:
@katspaugh, seems like both were CSS selector issues and these should now be addressed.

francovenica commented:
Checked all the notifications I could find and they look fine.

Regarding 
> One notable notification to check, which includes an onClick handler, is dispatched when the Safe requires an update. I removed some unnecessary code that is responsible for address book interaction notifications. They should still work as before as well.
I tried to use safes that are 1.0.0 or 1.1.1 and I didn't get any notification, not when accessing it nor when I upgrade them. So I'm not sure how to check this one

iamacook commented:
> Regarding
> 
> > One notable notification to check, which includes an onClick handler, is dispatched when the Safe requires an update. I removed some unnecessary code that is responsible for address book interaction notifications. They should still work as before as well.
> 
> I tried to use safes that are 1.0.0 or 1.1.1 and I didn't get any notification, not when accessing it nor when I upgrade them. So I'm not sure how to check this one

I've adjusted the action that shows the update notification. If you:

a) have an older Safe open and connect to an owner
b) switch to an older Safe whilst connected to an owner

the notification should be displayed. Clicking on it should take you to the details settings where you can update the Safe.

francovenica commented:
Ok, the notification is there.
I've checked that shows up when you load a safe with an outdated version (today is any < v1.3.0) or if you switch from one that is already up to date to one outdated. Also the notification only shows up if you are owner of the safe, so non-owners or disconnected users don't see the notification. Connecting to the app with the owner of an outdated safe while being having it loaded also shows the notification, which is correct.

Looks good to me
![image](https://user-images.githubusercontent.com/4503659/172595605-a5cc789a-510b-410f-b8b4-4697fd871b86.png)
